cardano-crypto-class/src/Cardano/Crypto/EllipticCurve/BLS12_381/C.hs

```diff
@@ -0,0 +1,618 @@
+  c_blst_affine_in_g = c_blst_p1_affine_in_g1
+
+  c_blst_generator = c_blst_p1_generator
+
+  c_blst_x_from_p = c_blst_x_from_p1
+  c_blst_y_from_p = c_blst_y_from_p1
+  c_blst_z_from_p = c_blst_z_from_p1
+
+  c_blst_x_from_affine = c_blst_x_from_affine1
+  c_blst_y_from_affine = c_blst_y_from_affine1
+  c_blst_p_is_equal = c_blst_p1_is_equal
+  c_blst_p_is_inf = c_blst_p1_is_inf
+
+instance BLS_Curve Curve2 where
+  c_blst_on_curve = c_blst_p2_on_curve
+
+  c_blst_add = c_blst_p2_add
+  c_blst_mult = c_blst_p2_mult
+  c_blst_cneg = c_blst_p2_cneg
+
+  c_blst_hash = c_blst_hash_to_g2
+  c_blst_compress = c_blst_p2_compress
+  c_blst_serialize = c_blst_p2_serialize
+  c_blst_uncompress = c_blst_p2_uncompress
+  c_blst_deserialize = c_blst_p2_deserialize
+
+  c_blst_in_g = c_blst_p2_in_g2
+  c_blst_to_affine = c_blst_p2_to_affine
+  c_blst_from_affine = c_blst_p2_from_affine
+  c_blst_affine_in_g = c_blst_p2_affine_in_g2
+
+  c_blst_generator = c_blst_p2_generator
+
+  c_blst_x_from_p = c_blst_x_from_p2
+  c_blst_y_from_p = c_blst_y_from_p2
+  c_blst_z_from_p = c_blst_z_from_p2
+
+  c_blst_x_from_affine = c_blst_x_from_affine2
+  c_blst_y_from_affine = c_blst_y_from_affine2
+
+  c_blst_p_is_equal = c_blst_p2_is_equal
+  c_blst_p_is_inf = c_blst_p2_is_inf
+
+---- Safe Scalar types / marshalling
+
+sizeScalar :: forall i. Num i => i
+sizeScalar = fromIntegral c_size_blst_scalar
+
+newtype Scalar = Scalar (ForeignPtr Void)
+
+withScalar :: Scalar -> (ScalarPtr -> IO a) -> IO a
+withScalar (Scalar p2) go = do
+  withForeignPtr p2 (go . ScalarPtr)
+
+withNewScalar :: (ScalarPtr -> IO a) -> IO (a, Scalar)
+withNewScalar go = do
+  p2 <- mallocForeignPtrBytes sizeScalar
+  x <- withForeignPtr p2 (go . ScalarPtr)
+  return (x, Scalar p2)
+
+withNewScalar_ :: (ScalarPtr -> IO a) -> IO a
+withNewScalar_ = fmap fst . withNewScalar
+
+withNewScalar' :: (ScalarPtr -> IO a) -> IO Scalar
+withNewScalar' = fmap snd . withNewScalar
+
+cloneScalar :: Scalar -> IO Scalar
+cloneScalar (Scalar a) = do
+  b <- mallocForeignPtrBytes sizeScalar
+  withForeignPtr a $ \ap ->
+    withForeignPtr b $ \bp ->
+      copyBytes bp ap sizeScalar
+  return (Scalar b)
+
+sizeFr :: forall i. Num i => i
+sizeFr = fromIntegral c_size_blst_fr
+
+newtype Fr = Fr (ForeignPtr Void)
+
+withFr :: Fr -> (FrPtr -> IO a) -> IO a
+withFr (Fr p2) go = do
+  withForeignPtr p2 (go . FrPtr)
+
+withNewFr :: (FrPtr -> IO a) -> IO (a, Fr)
+withNewFr go = do
+  p2 <- mallocForeignPtrBytes sizeFr
+  x <- withForeignPtr p2 (go . FrPtr)
+  return (x, Fr p2)
+
+withNewFr_ :: (FrPtr -> IO a) -> IO a
+withNewFr_ = fmap fst . withNewFr
+
+withNewFr' :: (FrPtr -> IO a) -> IO Fr
+withNewFr' = fmap snd . withNewFr
+
+cloneFr :: Fr -> IO Fr
+cloneFr (Fr a) = do
+  b <- mallocForeignPtrBytes sizeFr
+  withForeignPtr a $ \ap ->
+    withForeignPtr b $ \bp ->
+      copyBytes bp ap sizeFr
+  return (Fr b)
+
+---- Safe FP types / marshalling
+
+sizeFP :: forall i. Num i => i
+sizeFP = fromIntegral c_size_blst_scalar
```

Comment:
`c_size_blst_scalar` is 32, but the size of FP is 48. We need a `c_size_blst_fp` in `cbits/blst.h` which is then used here.

Reply:
Module removed.